### PR TITLE
perf(storage): simplify table watermark index

### PR DESCRIPTION
**Pull request by wenym1**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Avoid building per vnode table watermark index.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/storage/hummock_sdk/src/table_watermark.rs [hunk 1]
```diff
@@ -208,46 +186,74 @@ impl TableWatermarksIndex {
...
+                let _ = self.watermarks.pop_front();
+            } else {
+                break;
+            }
+        }
+
+        let mut index = 0;
+        while index < committed_watermark.watermarks.len() {
+            let (old_epoch, _) = &committed_watermark.watermarks[index];
+            if *old_epoch <= committed_epoch {
```
Comment (hzxa21) [inline src/storage/hummock_sdk/src/table_watermark.rs:242]: Can this condition ever be false? I thought the highest epoch for `committed_watermark` is `committed_epoch`.

File: src/storage/hummock_sdk/src/table_watermark.rs [hunk 2]
```diff
@@ -208,46 +186,74 @@ impl TableWatermarksIndex {
...
+        committed_epoch: HummockEpoch,
+    ) {
+        assert_eq!(self.watermark_direction, committed_watermark.direction);
+        assert!(self.committed_epoch <= committed_epoch);
+        if self.committed_epoch == committed_epoch {
+            return;
+        }
+        self.committed_epoch = committed_epoch;
+        // keep only watermark higher than committed epoch
+        while let Some((old_epoch, _)) = self.watermarks.front() {
```
Comment (hzxa21) [inline src/storage/hummock_sdk/src/table_watermark.rs:231]: Given that `watermarks` are sorted by epoch, will it be more efficient to use `binary_search` + `split_off` or `truncate` (need to swtich to `push_front`) here?

File: src/storage/hummock_sdk/src/table_watermark.rs [hunk 3]
```diff
@@ -208,46 +186,74 @@ impl TableWatermarksIndex {
...
+            if *old_epoch <= committed_epoch {
+                let _ = self.watermarks.pop_front();
+            } else {
+                break;
+            }
+        }
+
+        let mut index = 0;
+        while index < committed_watermark.watermarks.len() {
+            let (old_epoch, _) = &committed_watermark.watermarks[index];
```
Comment (hzxa21) [inline src/storage/hummock_sdk/src/table_watermark.rs:241]: I think we can use binary search here.

**Comment by wenym1** [comment]
Added benchmark.

Benchmark result is as followed. The test setting is, 500 epochs between safe epoch and committed epoch, and 500 staging epochs. Vnodes are partitioned into 16 parts.

This branch
```
new pinned version      time:   [37.867 µs 38.279 µs 38.689 µs]
                        change: [-100.000% -100.000% -100.000%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 7 outliers among 100 measurements (7.00%)
  3 (3.00%) high mild
  4 (4.00%) high severe

apply committed watermark
                        time:   [161.95 µs 167.16 µs 172.45 µs]
                        change: [-99.989% -99.989% -99.988%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 3 outliers among 100 measurements (3.00%)
  3 (3.00%) high mild

read latest watermark   time:   [2.4513 µs 2.5030 µs 2.6131 µs]
                        change: [+2.5368% +3.8159% +6.2119%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 3 outliers among 100 measurements (3.00%)
  1 (1.00%) high mild
  2 (2.00%) high severe

read committed watermark
                        time:   [81.068 µs 81.169 µs 81.271 µs]
                        change: [-0.2821% -0.0573% +0.1584%] (p = 0.61 > 0.05)
                        No change in performance detected.

```

main branch
```
read latest watermark   time:   [2.3970 µs 2.4011 µs 2.4056 µs]
                        change: [-85.105% -85.063% -85.021%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 3 outliers among 100 measurements (3.00%)
  3 (3.00%) high mild

read committed watermark
                        time:   [81.021 µs 81.195 µs 81.395 µs]
                        change: [+706.95% +709.75% +712.35%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 3 outliers among 100 measurements (3.00%)
  1 (1.00%) high mild
  2 (2.00%) high severe
```

The time for `build_index` and `apply_committed_watermarks` can be ignored in this branch compared to the main branch, because there is no index to maintain. The time in the main branch is so great that we had to reduce the sample size to be able to finish the test in reasonable time.

For query, the time to query the latest epoch is 85% lower than the main branch, and the time to query a middle epoch is 7 times the main branch. This is because the implementation is linearly searching from latest epoch to old epoch, and therefore is't biased for querying the watermark in the latest epoch.

In conclusion, compared to the main branch, this PR save the significant time to build and maintain the table watermark index. In terms of query, this branch is even faster than the main branch in querying the latest epoch. It only performs poorly in querying a mvcc epoch, which is not a common use case.

File: src/storage/hummock_sdk/src/table_watermark.rs [hunk 4]
```diff
@@ -79,23 +52,43 @@ impl TableWatermarksIndex {
...
 
     pub fn read_watermark(&self, vnode: VirtualNode, epoch: HummockEpoch) -> Option<Bytes> {
-        self.index.get(&vnode).and_then(|epoch_watermarks| {
-            epoch_watermarks
-                .upper_bound(Included(&epoch))
-                .value()
-                .cloned()
-        })
+        // iterate from new epoch to old epoch
+        for (watermark_epoch, vnode_watermark_list) in self.staging_watermarks.iter().rev().chain(
```
Comment (hzxa21) [inline src/storage/hummock_sdk/src/table_watermark.rs:77]: > the time to query a middle epoch is 7 times the main branch. This is because the implementation is linearly searching from latest epoch to old epoch, and therefore is't biased for querying the watermark in the latest epoch

This is a note only: I guess we can use binary search to make query from non-latest epoch faster but given that only internal state table contains table watermark and internal state table reads are all latest, we don't need to worry about that right now.

**Review by hzxa21 — APPROVED** [review]
LSTM! The micro-bench result looks good! Thanks for the quick fix.
